Detect mask usage for Across functions

SwapnilGaikwad opened the pull request:
Contributes to #101973


### Optimises the pattern XAcross(ConditionalSelect(mask, a, 0))
1. AddAcross(conditionalSelect(mask, a, 0))
 ✅ Signed → AddAcross(mask, a)
 ✅ Unsigned → AddAcross(mask, a)

2. AddSequentialAcross(a, conditionalSelect(mask, b, 0))
 ✅ Signed → AddSequentialAcross(mask, b)
 ✅ Unsigned → AddSequentialAcross(mask, b)

3. ❌ AddSequentialAcross(conditionalSelect(mask, a, 0), b)

4. OrAcross(conditionalSelect(mask, a, 0))
 ✅ Signed → OrAcross(mask, a)
 ✅ Unsigned → OrAcross(mask, a)

5. XorAcross(conditionalSelect(mask, a, 0))
 ✅ Signed → OrAcross(mask, a)
 ✅ Unsigned → OrAcross(mask, a)

6. MaxAcross(conditionalSelect(mask, a, 0))
 ❌ Signed
 ✅ Unsigned → MaxAcross(mask, a)

7. MaxNumberAcross(conditionalSelect(mask, a, 0))
 ❌ Signed
 No unsigned version available in the API. Takes only doubles or floats.

8. ❌ AndAcross(conditionalSelect(mask, a, 0))

9. ❌ MinAcross(conditionalSelect(mask, a, 0))

10. ❌ MinNumberAcross(conditionalSelect(mask, a, 0))




@kunalspathak @a74nh @dotnet/arm64-contrib

SwapnilGaikwad committed:
Detect mask usage for Across functions

Fixes: #101973

a74nh commented:
> 6\. MaxAcross(conditionalSelect(mask, a, 0))

This can probably be done with `MaxAcross(conditionalSelect(mask, a, MAX_INT))`. I'd leave that for a future PR.

a74nh commented on src/tests/JIT/opt/InstructionCombining/AcrossAndCselToAcross.cs:
```diff
@@ -0,0 +1,77 @@
+        //ARM64-FULL-LINE-NEXT: saddv {{d[0-9]+}}, {{p[0-9]+}}, {{z[0-9]+.b}}
+        var vec = Vector.Create(op2);
+        var mask = Vector.Create(op1);
+        return Sve.AddAcross(Sve.ConditionalSelect(mask, vec, Vector<sbyte>.Zero));
```
My concern with these optimizations is that users will not know about them and will end up writing non-optimal code. Is there somewhere we could document these?

kunalspathak replied:
yes, and there are lot of other non-sve non-arm64 scenarios where someone writes code produces sub-optimal code. The best we can do it put these things together in a blog post that we publish occasionally.

SwapnilGaikwad committed:
Fix build errors on x86

SwapnilGaikwad commented on src/tests/JIT/opt/InstructionCombining/AcrossAndCselToAcross.cs:
```diff
@@ -0,0 +1,77 @@
+public class AcrossAndCselToAcross
+{
+
+    [Theory]
```
Hi @kunalspathak, these are the SVE only tests and would fail on non-sve machines.
Is there any way to restrict them to run only when `Sve.IsSupported`, and ensure that the assembly is verified only then? 
I tried using `[TheoryConditional(nameof(IsSveSupported))]` attribute but couldn't get it work.

kunalspathak replied:
You can simply do something like this:

```csharp
    [Theory]
    [InlineData(1, 42)]
    [MethodImpl(MethodImplOptions.NoInlining)]
    public static Vector<long> addAcross_sbyte(sbyte op1, sbyte op2)
    {
        if (Sve.IsSupported)
        {
           //ARM64-FULL-LINE: cmpne {{p[0-9]+.b}}, {{p[0-9]+/z}}, {{z[0-9]+.b}}, #0
           //ARM64-FULL-LINE-NEXT: saddv {{d[0-9]+}}, {{p[0-9]+}}, {{z[0-9]+.b}}
           var vec = Vector.Create(op2);
           var mask = Vector.Create(op1);
           return Sve.AddAcross(Sve.ConditionalSelect(mask, vec, Vector<sbyte>.Zero));
        }
        else return Vector<long>.Zero;
    }
```

kunalspathak commented on src/coreclr/jit/gentree.cpp:
```diff
@@ -20533,8 +20533,9 @@ bool GenTree::isEmbeddedMaskingCompatibleHWIntrinsic() const
-               HWIntrinsicInfo::IsOptionalEmbeddedMaskedOperation(AsHWIntrinsic()->GetHWIntrinsicId());
+        return HWIntrinsicInfo::IsEmbeddedMaskedOperation(intrinsicId) ||
+               HWIntrinsicInfo::IsOptionalEmbeddedMaskedOperation(intrinsicId) ||
+               (intrinsicId == NI_Sve_ConditionalSelect);
```
should this condition instead be part of one of `IsEmbeddedMaskedOperation` or `IsOptionalEmbeddedMaskedOperation` to make sure there are no missing scenarios?

kunalspathak replied:
the summar docs of this method should change and include arm64 portion as well.

kunalspathak replied:
looks like `isEmbeddedMaskingCompatibleHWIntrinsic` is just used at 2 places - `canBeContained()` and `ContainCheckHWIntrinsic`. For the later, we make sure that it is not a `CndSel `, so essentially we just need this condition for `canBeContained()`. Wondering if you can add this condition somewhere like `getContainedCselOperand()` itself?

SwapnilGaikwad replied:
Moved the containable check to `isContainableHWIntrinsic`. Apparently, csel is the first Arm64 one there.

kunalspathak commented on src/coreclr/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -1069,6 +1080,23 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
                     {
                         GetEmitter()->emitIns_R_R(ins, emitSize, targetReg, op1Reg, opt);
                     }
+                    else if (HWIntrinsicInfo::IsEmbeddedMaskedOperation(intrin.id) && intrin.op2->isContained())
```
I didn't understand which scenario this is trying to handle?

SwapnilGaikwad replied:
This is for `AddSequentialAcross` (`fadda`). This is the only intrinsic amongst the XAcross we are optimising where the optimisation is possible when the csel is at op2 and it meets other conditions ( zero as the third op of csel).
In the above list in the description, we can optimise item 2. (csel as second op of across) but not 3. (csel as the first operand of across).
For item 3, we get the wrong result if we do the current optimisation.

kunalspathak replied:
can you show the codegen for 2 and 3?

SwapnilGaikwad replied:
For case 2 (with optimisation)

<details>
<summary> Code example </summary>

```C#
public static Vector<double> Sve_AddSequentialAcross1()
{
    var vec1 = Vector.Create((double)42);
    var vec2 = Vector.Create((double)43);
    var mask = Vector.Create((double)1);
    return Sve.AddSequentialAcross(vec2, Sve.ConditionalSelect(mask, vec1, Vector<double>.Zero));
}

```
</details>


```
...
G_M19924_IG02:  ;; offset=0x0018
            ldr     d0, [@RWD00]
            stp     d0, d0, [fp, #0x30]	// [V04 tmp2], [V04 tmp2+0x08]
            ldr     q0, [fp, #0x30]	// [V04 tmp2]
            ldr     d16, [@RWD08]
            stp     d16, d16, [fp, #0x20]	// [V09 tmp7], [V09 tmp7+0x08]
            ldr     q16, [fp, #0x20]	// [V09 tmp7]
            fmov    d17, #1.0000
            stp     d17, d17, [fp, #0x10]	// [V14 tmp12], [V14 tmp12+0x08]
            ldr     q17, [fp, #0x10]	// [V14 tmp12]
            ptrue   p0.d
            cmpne   p0.d, p0/z, z17.d, #0
            fadda   v16.d, p0, v16.d, z0.d
            mov     v0.16b, v16.16b

G_M19924_IG03:  ;; offset=0x004C
            ldp     fp, lr, [sp], #0x40
            ret     lr
```


For case 3 (no optimisation)

<details>
<summary> Code example </summary>

```C#
public static Vector<double> Sve_AddSequentialAcross1()
{
    var vec1 = Vector.Create((double)42);
    var vec2 = Vector.Create((double)43);
    var mask = Vector.Create((double)1);
    return Sve.AddSequentialAcross(Sve.ConditionalSelect(mask, vec1, Vector<double>.Zero), vec2);
}

```
</details>

```
...
G_M38135_IG02:  ;; offset=0x0018
            ldr     d0, [@RWD00]
            stp     d0, d0, [fp, #0x30]	// [V04 tmp2], [V04 tmp2+0x08]
            ldr     q0, [fp, #0x30]	// [V04 tmp2]
            ldr     d16, [@RWD08]
            stp     d16, d16, [fp, #0x20]	// [V09 tmp7], [V09 tmp7+0x08]
            ldr     q16, [fp, #0x20]	// [V09 tmp7]
            fmov    d17, #1.0000
            stp     d17, d17, [fp, #0x10]	// [V14 tmp12], [V14 tmp12+0x08]
            ldr     q17, [fp, #0x10]	// [V14 tmp12]
            ptrue   p0.d
            cmpne   p0.d, p0/z, z17.d, #0
            movi    v17.4s, #0
            sel     z0.d, p0, z0.d, z17.d
            ptrue   p0.d
            fadda   v0.d, p0, v0.d, z16.d

G_M38135_IG03:  ;; offset=0x0054
            ldp     fp, lr, [sp], #0x40
            ret     lr
```

kunalspathak replied:
For case 3, there are multiple things I am spotting odd:

1. `movi v17.4s, #0` : We want to move 0 to scalable register I assume because that's what we are using in next `sel` instruction, so shouldn't this be `fmov z17.s, #0`?
2. If I see `AddSequentialAcross(initial, value)` and my understanding is `initial` is the source / destination SIMD register, but above, when we do `fadda`, we are giving it `v0.d` as source register which is populated using `sel` in scalable register. When we support VL agnostic, won't we lose data? I am thinking this falls under the same category as [LASTB](https://docsmirror.github.io/A64/2023-06/lastb_v_p_z.html) where destination is SIMD & FP register instead of scalable register. (reference: https://github.com/dotnet/runtime/pull/103847#discussion_r1658060716). If yes, then we should rethink about this API or perhaps, if someone uses `ConditionalSelect()`, here we will just clobber the data and operate on SIMD registers.

SwapnilGaikwad replied:
In case 1., it seems to be optimised `fmov` to `movi` as we are moving zero.
Case 2., you're right. We may need to update the API calls to reflect that the initial and return value is of type `Vector128` while the op2 is of type `Vector<T>`.

kunalspathak replied:
@tannergooding - what is your thoughts on this?

kunalspathak commented on src/tests/JIT/HardwareIntrinsics/Arm/Shared/_SveMinimalUnaryOpTestTemplate.template:
```diff
@@ -294,6 +297,17 @@ namespace JIT.HardwareIntrinsics.Arm
+            TestLibrary.TestFramework.BeginScenario("RunScenario_AbsorbConditionalSelectOp - zero in falseValue");
+            var mask = Vector.As<{RetBaseType}, {Op1BaseType}>(_mask);
+            var zeroVec = {Op1VectorType}<{Op1BaseType}>.Zero;
+            var result = {Isa}.{Method}(Sve.ConditionalSelect(mask, _fld1, zeroVec));
```
we do not support predicate version of `sel` yet, right? otherwise something like `Sve.ConditionalSelect(mask, mask1, mask2)` could be one more test scenario.

SwapnilGaikwad replied:
True. We don't support the predicate version of sel yet.

kunalspathak commented on src/tests/JIT/opt/InstructionCombining/AcrossAndCselToAcross.cs:
```diff
@@ -0,0 +1,77 @@
+        var vec1 = Vector.Create(op2);
+        var vec2 = Vector.Create(op3);
+        var mask = Vector.Create(op1);
+        return Sve.AddSequentialAcross(vec1, Sve.ConditionalSelect(mask, vec2, Vector<double>.Zero));
```
can we also add a scenario for `AllBitsSet`, `One` and `vec2` or something for completion?

SwapnilGaikwad replied:
Updated one of the tests to take the value with all 1s as input. 👍

kunalspathak reviewed (commented):
thanks for lighting up this optimization.

SwapnilGaikwad committed:
Refactor containable check for csel

SwapnilGaikwad committed:
Merge main

kunalspathak commented:
also, seems there are failures in new test you added:

```
__tmp0_AcrossAndCselToAcross.cs:24:11: error: ARM64: expected string not found in input
//ARM64: {{^ *}} cmpne {{p[0-9]+.b}}, {{p[0-9]+/z}}, {{z[0-9]+.b}}, #0{{$}}
^
__jit_disasm.out:16:64: note: scanning from here
; BEGIN METHOD AcrossAndCselToAcross:addAcross_sbyte(byte,byte):System.Numerics.Vector`1[long]
^
__jit_disasm.out:26:2: note: possible intended match here
ldp fp, lr, [sp], #0x10
^
```

SwapnilGaikwad commented:
> also, seems there are failures in new test you added:
> 
> ```
> __tmp0_AcrossAndCselToAcross.cs:24:11: error: ARM64: expected string not found in input
> //ARM64: {{^ *}} cmpne {{p[0-9]+.b}}, {{p[0-9]+/z}}, {{z[0-9]+.b}}, #0{{$}}
> ^
> __jit_disasm.out:16:64: note: scanning from here
> ; BEGIN METHOD AcrossAndCselToAcross:addAcross_sbyte(byte,byte):System.Numerics.Vector`1[long]
> ^
> __jit_disasm.out:26:2: note: possible intended match here
> ldp fp, lr, [sp], #0x10
> ^
> ```

This is the issue that assembly checks are not scoped. They are tested for the entire method. Unfortunately, that mean when the test is skipped (`!Sve.IsSupported`), the assembly won't match. 
Do you have any suggestions to work the assembly check on non-sve machines?

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Remove optimisation of AddSequentialAcross and assembly checks

SwapnilGaikwad commented:
Dropped the optimisation for `AddSequentialAcross` and created an issue #109643 to track it.
Also removed the assembly checks as they cannot be tested in absence of an SVE enabled machine in the CI. Created an issue to add assembly check #109644.

kunalspathak commented on src/coreclr/jit/lowerarmarch.cpp:
```diff
@@ -1881,17 +1881,32 @@ GenTree* Lowering::LowerHWIntrinsic(GenTreeHWIntrinsic* node)
         DISPTREERANGE(BlockRange(), node);
         JITDUMP("\n");
 
+        // Use lastOp to verify if it's a ConditionlSelectNode. For AddSequentialAcross intrinsic, we can absorb the
```
can you remove the reference for `AddSequentialAcross`?

SwapnilGaikwad replied:
Done.

kunalspathak commented on src/coreclr/jit/lowerarmarch.cpp:
```diff
@@ -3655,6 +3670,58 @@ bool Lowering::TryLowerNegToMulLongOp(GenTreeOp* op, GenTree** next)
+    {
+        switch (parentNode->GetHWIntrinsicId())
+        {
+            case NI_Sve_AddAcross:
```
Remind me but why we cannot support these?
- AndAcross
- signed/unsigned MinAcross
- signed MaxAcross

SwapnilGaikwad replied:
Inactive elements are treated as zero so absorbing the mask may yield an incorrect result.

kunalspathak replied:
ok, but they can be optimized if `mask == AllBitsOne` for example?

SwapnilGaikwad replied:
A quick update, we optimise `Csel(AllTrue, op1, op2) => op1` already [here](https://github.com/dotnet/runtime/blob/6aaadea3a888defc631456dc394c95727a876f31/src/coreclr/jit/lowerarmarch.cpp#L4165-L4207) so don't need to do it for XAcross scenarios.

kunalspathak commented on src/tests/JIT/HardwareIntrinsics/Arm/Shared/_SveMinimalUnaryOpTestTemplate.template:
```diff
@@ -294,6 +297,17 @@ namespace JIT.HardwareIntrinsics.Arm
+            TestLibrary.TestFramework.BeginScenario("RunScenario_AbsorbConditionalSelectOp - zero in falseValue");
+            var mask = Vector.As<{RetBaseType}, {Op1BaseType}>(_mask);
+            var zeroVec = {Op1VectorType}<{Op1BaseType}>.Zero;
+            var result = {Isa}.{Method}(Sve.ConditionalSelect(mask, _fld1, zeroVec));
```
do we have coverage for `Sve.AddAcross(Sve.ConditionalSelect(AllOne, _fld1, zero))` or some more combination like that?

SwapnilGaikwad replied:
We don't at the moment but happy to add if you suggest so 🙂 .

kunalspathak replied:
worth adding few to support https://github.com/dotnet/runtime/pull/109218#discussion_r1837063846 scenario.

SwapnilGaikwad replied:
Done.

kunalspathak reviewed (commented):
Added some questions. Can you also confirm if all the tests pass on stress?

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Add more tests with different mask values

SwapnilGaikwad committed:
Avoid containing csel that's already containing an embedded op

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Add test for the Fuzzlyn reported failure

SwapnilGaikwad committed:
Fix test build failures

SwapnilGaikwad committed:
Merge main

kunalspathak commented on src/coreclr/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -1053,6 +1053,17 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+                        assert(cselIntrin.id == NI_Sve_ConditionalSelect);
+                        regNumber maskReg = cselIntrin.op1->GetRegNum();
+                        op1Reg            = cselIntrin.op2->GetRegNum();
+                        GetEmitter()->emitIns_R_R_R(ins, emitSize, targetReg, maskReg, op1Reg, opt);
```
can we add `assert(cselIntrin.op3->IsVectorZero())`;

kunalspathak reviewed (approved):
I added some nits that you can do in a follow-up PR.

kunalspathak commented:
/ba-g timeout infra issue.
